In [ ]:
!pip install rag-evaluator

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 51.8/51.8 kB 4.9 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.1/61.1 kB 6.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 104.0/104.0 kB 9.8 MB/s eta 0:00:00
  Created wheel for rouge-score: filename=rouge_score-0.1.2-py3-none-any.whl size=24935 sha256=45bc21c16f3f39a33dfb20385f05f1488150996179de2f322da8487d28760531
  Stored in directory: /root/.cache/pip/wheels/5f/dd/89/461065a73be61a532ff8599a28e9beef17985c9e9c31e541b4
Successfully built rouge-score


1.First LLM MODEL flan-t5-large

https://huggingface.co/google/flan-t5-large


In [ ]:
import math
from rag_evaluator import RAGEvaluator

# Initialize the evaluator
evaluator = RAGEvaluator()

# QA data with questions, responses, and reference answers
qa_data = [
    {
        "question": "What are the general immigration requirements for newcomers in Germany?",
        "response": "• Proof of sufficient financial means • Health insurance • Basic knowledge of the German language • Third-country national must sign a declaration that he is aware that the permit can be refused and expulsion can be ordered if false information is provided • Other documents necessary to prove the purpose of immigration",
        "reference": "Proof of sufficient financial means, health insurance, basic knowledge of the German language, and other necessary documents to prove the purpose of immigration."
    },
    {
        "question": "How long does the visa application process take in Germany?",
        "response": "three months",
        "reference": "The visa application process typically takes three months."
    },
    {
        "question": "What are the family reunification policies in Germany?",
        "response": "Spouses and minor children of German citizens may be issued a residence permit for the purpose of family reunification, as may the parents of minor unmarried Germans for the purpose of care",
        "reference": "Spouses and minor children of German citizens may be issued a residence permit for family reunification, as may parents of minor unmarried Germans for the purpose of care."
    },
    {
        "question": "Are there any specific language requirements to migrate to Germany?",
        "response": "Knowledge of the German language is not required prior to entry",
        "reference": "Knowledge of the German language is not required prior to entry."
    },
    {
        "question": "What types of work visas are available in Germany?",
        "response": "seasonal work",
        "reference": "Seasonal work visas are available in Germany."
    },
    {
        "question": "Can I apply for permanent residency in Germany after a certain period?",
        "response": "2.3.2.2 Permanent Residence",
        "reference": "Information regarding permanent residency eligibility varies based on the applicant's situation, such as employment or study duration."
    },
    {
        "question": "What is the capital of Pakistan?",
        "response": "Pakistan",
        "reference": "The capital of Pakistan is Islamabad."
    }
]

# Print table header
print(f"{'Q#':<5} {'Question':<60} {'BLEU':<8} {'ROUGE-1':<8} {'BERT P':<8} {'BERT R':<8} {'BERT F1':<8} {'Perplexity':<12} {'Diversity':<10} {'Racial Bias':<12}")
print("="*110)

# Evaluate and print the results
for idx, qa in enumerate(qa_data, 1):
    question = qa["question"]
    response = qa["response"]
    reference = qa["reference"]

    # Evaluate using the RAGEvaluator
    metrics = evaluator.evaluate_all(question, response, reference)

    # Extract the metrics, handling NaN gracefully and rounding values to 2 decimal places
    bleu = round(metrics.get('BLEU', float('nan')), 2) if metrics.get('BLEU', float('nan')) != float('nan') else 'N/A'
    rouge_1 = round(metrics.get('ROUGE-1', float('nan')), 2) if metrics.get('ROUGE-1', float('nan')) != float('nan') else 'N/A'
    bert_p = round(metrics.get('BERT P', float('nan')), 2) if metrics.get('BERT P', float('nan')) != float('nan') else 'N/A'
    bert_r = round(metrics.get('BERT R', float('nan')), 2) if metrics.get('BERT R', float('nan')) != float('nan') else 'N/A'
    bert_f1 = round(metrics.get('BERT F1', float('nan')), 2) if metrics.get('BERT F1', float('nan')) != float('nan') else 'N/A'
    perplexity = round(metrics.get('Perplexity', float('nan')), 2) if metrics.get('Perplexity', float('nan')) != float('nan') else 'N/A'
    diversity = round(metrics.get('Diversity', float('nan')), 2) if metrics.get('Diversity', float('nan')) != float('nan') else 'N/A'
    racial_bias = round(metrics.get('Racial Bias', float('nan')), 2) if metrics.get('Racial Bias', float('nan')) != float('nan') else 'N/A'

    # Print results in tabular format
    print(f"{idx:<5} {question[:60]:<60} {bleu:<8} {rouge_1:<8} {bert_p:<8} {bert_r:<8} {bert_f1:<8} {perplexity:<12} {diversity:<10} {racial_bias:<12}")


Hardware accelerator e.g. GPU is available in the environment, but no `device` argument is passed to the `Pipeline` object. Model will be on CPU.
Failed to determine 'entailment' label id from the label2id mapping in the model config. Setting to -1. Define a descriptive label2id mapping in the model config to ensure correct outputs.


Q#    Question                                                     BLEU     ROUGE-1  BERT P   BERT R   BERT F1  Perplexity   Diversity  Racial Bias 
1     What are the general immigration requirements for newcomers  22.56    0.64     0.73     0.85     0.79     60.43        0.96       0.5         
2     How long does the visa application process take in Germany?  0.0      0.4      0.78     0.69     0.73     1205.35      0.5        0.47        
3     What are the family reunification policies in Germany?       77.09    0.93     0.94     0.95     0.95     27.58        0.88       0.46        
4     Are there any specific language requirements to migrate to G 91.31    1.0      0.97     0.96     0.97     30.22        0.91       0.47        
5     What types of work visas are available in Germany?           0.0      0.44     0.82     0.75     0.79     13952.01     0.5        0.45        
6     Can I apply for permanent residency in Germany after a certa 0.0      0.17     0.62     0.61     0.6

**Interpretation**

Best Performing: Question 3 and 4 (high BLEU, ROUGE, and BERT scores, low perplexity, high diversity)

Worst Performing: Question 2, 5, 6, and 7 (very high perplexity, low BLEU, and low semantic similarity)

Diversity: Questions 1, 3, and 4 have high diversity, while Question 7 has very low diversity.

Racial Bias: Moderate bias is present across most outputs, especially for Questions 1, 2, and 5.

2. Second LLM Model bart-large

https://huggingface.co/facebook/bart-large

In [ ]:
from rag_evaluator import RAGEvaluator
import math

# Initialize the evaluator
evaluator = RAGEvaluator()

# Define a list of question-answer pairs
qa_data = [
    {
        "question": "What are the general immigration requirements for newcomers in Germany?",
        "response": ("539Laws for Legal Immigration in the 27 EU Member States¯¯¯¯¯¯¯¯Authority competent "
                     "to issue residence permitConditions Applicable for Residence¯¯¯¯¯¯¯¯Regardless of Immigration "
                     "CategoryGrounds for Rejection Appeal Rights Fees503GermanyRegional Aliens¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯━..."),
        "reference": ("Proof of sufficient financial means, health insurance, basic knowledge of the German language, "
                     "and other necessary documents to prove the purpose of immigration.")
    },
    {
        "question": "What are the family reunification policies in Germany?",
        "response": ("286 Family reunification for German citizens. Spouses and minor children of German citizens may be issued a "
                     "residence permit for the purpose of family reunification, as may the parents of minor unmarried Germans for the purpose "
                     "of care (Residence Act, Art. 28(5)). The residence permit shall be prolonged as long as the long-term family relationship persists. "
                     "Foreign family members of Germans have the right to be employed or self-employed without examination of the labour market or other restrictions "
                     "(residence act, art. 28). The right of residence for spouses of Germans may be extended to spouses of non-German citizens if the spouse of a German citizen "
                     "is a citizen of a foreign country. This right is based on the principle of family respect for the fundamental principles recognized by the laws of the Republic "
                     "(primarily the principle in the Constitution of the Federal Republic and the principles of equality between genders and the rejection of discrimination on the basis of origin)."),
        "reference": ("Spouses and minor children of German citizens may be issued a residence permit for family reunification, as may parents of minor unmarried Germans for the purpose of care.")
    },
    {
        "question": "Are there any specific language requirements to migrate to Germany?",
        "response": ("Knowledge of the German language is not required prior to entry; however, this may change. 85 For countries that apply language requirements as a general condition and have national permanent residence permits, i.e., Austria and Germany. "
                     "The applicant does not, however, have to prove knowledge of the host country’s history. 86 Exceptions are made for persons not aged between 18 and 65, former members of the armed forces, and for persons who have already been living in Germany for a long period of time. "
                     "Currently, in order to obtain a permanent residence permit, the applicant must show a basic knowledge of the German language. At present, immigrants are expected to make the effort to learn the German language only after their entry, to which end language courses are offered. "
                     "Immigrants with insufficient language skills can be required to attend these courses (Residence Act, Art. 44(a)). In practice, the courses are used most frequently by the first generation of immigrants."),
        "reference": "Knowledge of the German language is not required prior to entry."
    },
    {
        "question": "What types of work visas are available in Germany?",
        "response": ("Including: seasonal work; au-pair occupation; domestic help in households with persons in need of care; and social work. Also subject to permissive regulations for entry and stay are those individuals belonging to German ethnic groups in Eastern European countries or those States formerly part of the Soviet Union, persons displaced in World War II and their descendants, and foreigners who lived in the federal territory as minors and who have a right to return after a durable departure (Residence Act, Art. 37). "
                     "General rules apply for 'switching' of immigration category. Permit is issued a maximum of 6 months within 1 year. Procedural requirements are similar to the general rules."),
        "reference": "Seasonal work visas are available in Germany."
    },
    {
        "question": "Can I apply for permanent residency in Germany after a certain period?",
        "response": ("The time of the applicant’s studies or vocational training can be counted for up to half of the required stay time. "
                     "The maximum length of stay varies according to the purpose of stay. For example, a temporary residence permit may be given within a period of two months to an employee sent temporarily for the supply of services (Employment Regulation, Art. 31), or up to three months within a month for the provision of services to an immigrant sent temporarily to Germany for a short period of time. "
                     "The duration of the stay is determined by the duration of an economic activity. Generally, renewal of a temporary residency permit is subject to the same conditions that apply to its first issuance (Residence Act, Art. No. 11)."),
        "reference": "Permanent residency can be applied for after fulfilling certain conditions like time of stay, studies, or economic activity duration in Germany."
    }
]

# Print table header
print(f"{'Q#':<5} {'Question':<60} {'BLEU':<8} {'ROUGE-1':<8} {'BERT P':<8} {'BERT R':<8} {'BERT F1':<8} {'Perplexity':<12} {'Diversity':<10} {'Racial Bias':<12}")
print("="*110)

# Evaluate and print the results
for idx, qa in enumerate(qa_data, 1):
    question = qa["question"]
    response = qa["response"]
    reference = qa["reference"]

    # Evaluate using the RAGEvaluator
    metrics = evaluator.evaluate_all(question, response, reference)

    # Extract the metrics, handling NaN gracefully and rounding values to 2 decimal places
    bleu = round(metrics.get('BLEU', float('nan')), 2) if metrics.get('BLEU', float('nan')) != float('nan') else 'N/A'
    rouge_1 = round(metrics.get('ROUGE-1', float('nan')), 2) if metrics.get('ROUGE-1', float('nan')) != float('nan') else 'N/A'
    bert_p = round(metrics.get('BERT P', float('nan')), 2) if metrics.get('BERT P', float('nan')) != float('nan') else 'N/A'
    bert_r = round(metrics.get('BERT R', float('nan')), 2) if metrics.get('BERT R', float('nan')) != float('nan') else 'N/A'
    bert_f1 = round(metrics.get('BERT F1', float('nan')), 2) if metrics.get('BERT F1', float('nan')) != float('nan') else 'N/A'
    perplexity = round(metrics.get('Perplexity', float('nan')), 2) if metrics.get('Perplexity', float('nan')) != float('nan') else 'N/A'
    diversity = round(metrics.get('Diversity', float('nan')), 2) if metrics.get('Diversity', float('nan')) != float('nan') else 'N/A'
    racial_bias = round(metrics.get('Racial Bias', float('nan')), 2) if metrics.get('Racial Bias', float('nan')) != float('nan') else 'N/A'

    # Print results in tabular format
    print(f"{idx:<5} {question[:60]:<60} {bleu:<8} {rouge_1:<8} {bert_p:<8} {bert_r:<8} {bert_f1:<8} {perplexity:<12} {diversity:<10} {racial_bias:<12}")

Hardware accelerator e.g. GPU is available in the environment, but no `device` argument is passed to the `Pipeline` object. Model will be on CPU.
Failed to determine 'entailment' label id from the label2id mapping in the model config. Setting to -1. Define a descriptive label2id mapping in the model config to ensure correct outputs.


Q#    Question                                                     BLEU     ROUGE-1  BERT P   BERT R   BERT F1  Perplexity   Diversity  Racial Bias 
1     What are the general immigration requirements for newcomers  1.76     0.15     0.56     0.64     0.6      298.0        0.96       0.49        
2     What are the family reunification policies in Germany?       13.71    0.29     0.7      0.88     0.78     16.46        0.9        0.48        
3     Are there any specific language requirements to migrate to G 5.11     0.13     0.59     0.83     0.69     20.14        0.93       0.49        
4     What types of work visas are available in Germany?           0.41     0.07     0.59     0.73     0.65     58.07        0.99       0.49        
5     Can I apply for permanent residency in Germany after a certa 1.28     0.2      0.63     0.74     0.68     22.45        0.89       0.46        


**Interpretation**

Best Performing: Question 2 (good BLEU, ROUGE, BERT scores, low perplexity, high diversity)

Worst Performing: Question 4 (very low BLEU and ROUGE scores, moderate perplexity)

Diversity: All questions show very high diversity (ranging from 0.89 to 0.99).

Racial Bias: All questions show moderate bias (ranging from 0.46 to 0.49).

Perplexity: Question 2 and Question 5 have the lowest perplexity (better coherence), while Question 1 and Question 4 have higher perplexity (less coherent).

**CONCLUSION**

Model 1 is performing better overall due to its higher BLEU, ROUGE-1, and BERT scores, lower perplexity, and better semantic quality. Model 2 has significantly lower performance across these metrics, suggesting it is less effective.